documentation/specs/Table.md

```diff
@@ -0,0 +1,354 @@
+# `Table` Component Specification
+
+## Overview
+
+A `Table` is an interactive data grid used for working with a large collection of data in a scannable way.
+
+---
+
+## Design
+
+`Table` will use the `useTable` hooks from React Aria for managing display, selection, and sort. React Aria doesn't support expansion capabilities out of the box, so an additional layer on top of Aria's will need to be built to support expansion.
+
+Data will be passed to the `Table` component through the `columns` and `rows` prop. This allows us to support Aria's dynamic collections API through a single prop interface. Custom rendering within parts of the table will be supported through [render props](https://react.dev/reference/react/cloneElement#passing-data-with-a-render-prop).
+
+`columns` and `rows` will be arrays of objects and must follow a certain signature. `columns` should define the column structure for the table and must contain a unique `key`. `rows` should contain the content of the table and must contain a unique `key` along with properties for each `key` defined in the `columns` definition. The remainder of the data structure can be arbitrary. Column and row keys will be passed to render prop functions to control their rendering.
+
+_`columns` definition example_:
+
+```tsx
+const columns = [
+  { key: "name", name: "Name" },
+  { key: "type", name: "Type" },
+  { key: "date", name: "Date Modified" },
+] as const;
+```
+
+_`rows` definition example_:
+
+```tsx
+const rows = [
+  { key: 1, name: "filename_1", type: "js", date: new Date(2023, 1, 1) },
+  { key: 2, name: "filename_2", type: "js", date: new Date(2023, 1, 2) },
+] as const;
+```
+
+### API
+
+```ts
+type TableProps<C extends Column> = AriaLabelingProps & {
+  /** Use columns and rows to specify Table content. */
+  children?: never;
+
+  /** Columns for the table. */
+  columns: readonly C[];
+
+  /** The initial expanded key in the collection (uncontrolled). */
+  defaultExpandedKey?: Key;
+
+  /** The initial selected keys in the collection (uncontrolled). */
+  defaultSelectedKeys?: "all" | Iterable<Key>;
+
+  /** The currently expanded key in the collection (controlled). */
+  expandedKey?: Key;
+
+  /** Whether or not row expansion is enabled. */
+  hasExpandableRows?: boolean;
```

Comment:
**Question:** Are you thinking there's a case where we would provide any of the other expandable props (`expandedKey`, `onExpandedChange`, `renderExpandedRow` but want to disable expandability?

Reply:
no i don't imagine that being a case. and i went back and forth on this option as a mechanism for expressed intentionality of row expansion, but i kinda tend to think it's superfluous with the other props